File: AWS_GUI_Dev/BAWGUI/Model/PMUWithSamplingRate.vb
Imports BAWGUI

Public Class PMUWithSamplingRate
    Inherits ViewModelBase
    Public Sub New(pmu As String, rate As Integer)
        _pmu = pmu
        _samplingRate = rate
    End Sub
    Public Sub New()

    End Sub
    Private _pmu As String
    Public Property PMU As String
        Get
            Return _pmu
        End Get
        Set(ByVal value As String)
            _pmu = value
            OnPropertyChanged()
        End Set
    End Property
    Private _samplingRate As Integer
    Public Property SamplingRate As Integer
        Get
            Return _samplingRate
        End Get
        Set(ByVal value As Integer)
            _samplingRate = value
            OnPropertyChanged()
        End Set
    End Property
    Public Overrides Function Equals(obj As Object) As Boolean
        If obj Is Nothing OrElse Not Me.GetType Is obj.GetType Then
            Return False
        End If
        Dim p As PMUWithSamplingRate = CType(obj, PMUWithSamplingRate)
        Return Me.PMU = p.PMU ' AndAlso Me.SamplingRate = p.SamplingRate
    End Function
    Public Overrides Function GetHashCode() As Integer
        'If item Is Nothing Then Return 0
        'Dim hashItemName = If(item.PMU Is Nothing, 0, item.PMU.GetHashCode())
        'Dim hashRate = item.SamplingRate.GetHashCode()
        Return PMU.GetHashCode() ' Xor SamplingRate.GetHashCode()
    End Function
    Public Shared Operator =(ByVal x As PMUWithSamplingRate, ByVal y As PMUWithSamplingRate) As Boolean
        Return x.PMU = y.PMU ' AndAlso x.SamplingRate = y.SamplingRate
    End Operator
    Public Shared Operator <>(ByVal x As PMUWithSamplingRate, ByVal y As PMUWithSamplingRate) As Boolean
        Return x.PMU <> y.PMU ' OrElse x.SamplingRate <> y.SamplingRate
    End Operator
End Class

Public Class PMUWithSamplingRateComparer
    Implements IEqualityComparer(Of PMUWithSamplingRate)

    Public Overloads Function Equals(ByVal x As PMUWithSamplingRate, ByVal y As PMUWithSamplingRate) As Boolean Implements IEqualityComparer(Of PMUWithSamplingRate).Equals
        If x Is y Then Return True
        If x Is Nothing OrElse y Is Nothing Then Return False
        Return (x.PMU = y.PMU) ' AndAlso (x.SamplingRate = y.SamplingRate)
    End Function
    Public Overloads Function GetHashCode(ByVal item As PMUWithSamplingRate) As Integer Implements IEqualityComparer(Of PMUWithSamplingRate).GetHashCode
        If item Is Nothing Then Return 0
        Dim hashItemName = If(item.PMU Is Nothing, 0, item.PMU.GetHashCode())
        'Dim hashRate = item.SamplingRate.GetHashCode()
        Return hashItemName ' Xor hashRate
    End Function
End Class

File: AWS_GUI_Dev/BAWGUI.SignalManagement/ViewModels/PMUWithSamplingRate.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using BAWGUI.Utilities;
using Microsoft.VisualBasic;

namespace BAWGUI.SignalManagement.ViewModels
{
    public class PMUWithSamplingRate : ViewModelBase
    {
        public PMUWithSamplingRate(string pmu, int rate)
        {
            _pmu = pmu;
            _samplingRate = rate;
        }
        public PMUWithSamplingRate()
        {
        }
        private string _pmu;
        public string PMU
        {
            get
            {
                return _pmu;
            }
            set
            {
                _pmu = value;
                OnPropertyChanged();
            }
        }
        private int _samplingRate;
        public int SamplingRate
        {
            get
            {
                return _samplingRate;
            }
            set
            {
                _samplingRate = value;
                OnPropertyChanged();
            }
        }
        public override bool Equals(object obj)
        {
            if (obj == null || this.GetType() != obj.GetType())
                return false;
            PMUWithSamplingRate p = (PMUWithSamplingRate)obj;
            return this.PMU == p.PMU; // AndAlso Me.SamplingRate = p.SamplingRate
        }
        public override int GetHashCode()
        {
            // If item Is Nothing Then Return 0
            // Dim hashItemName = If(item.PMU Is Nothing, 0, item.PMU.GetHashCode())
            // Dim hashRate = item.SamplingRate.GetHashCode()
            return PMU.GetHashCode(); // Xor SamplingRate.GetHashCode()
        }
        public static bool operator ==(PMUWithSamplingRate x, PMUWithSamplingRate y)
        {
            return x.PMU == y.PMU; // AndAlso x.SamplingRate = y.SamplingRate
        }
        public static bool operator !=(PMUWithSamplingRate x, PMUWithSamplingRate y)
        {
            return x.PMU != y.PMU; // OrElse x.SamplingRate <> y.SamplingRate
        }
    }

    public class PMUWithSamplingRateComparer : IEqualityComparer<PMUWithSamplingRate>
    {
        public new bool Equals(PMUWithSamplingRate x, PMUWithSamplingRate y)
        {
            if (x == y)
                return true;
            if (x == null || y == null)
                return false;
            return (x.PMU == y.PMU); // AndAlso (x.SamplingRate = y.SamplingRate)
        }
        public new int GetHashCode(PMUWithSamplingRate item)
        {
            if (item == null)
                return 0;
            var hashItemName = item.PMU == null ? 0 : item.PMU.GetHashCode();
            // Dim hashRate = item.SamplingRate.GetHashCode()
            return hashItemName; // Xor hashRate
        }
    }
}
